compiler/src/iree/compiler/Codegen/ExternalInterfaces/Utils.cpp

```diff
@@ -120,4 +121,288 @@ void storeNamedAttrIfPresent(SmallVectorImpl<NamedAttribute> &config,
+      SmallVector<int64_t> transposePerm =
+          llvm::to_vector(llvm::seq<int64_t>(0, inRank));
+      for (auto perm : materializeEncodingInfo.swizzle->permutation) {
+        transposePerm.push_back(inRank + perm);
+      }
+      applyPermutationToVector(packedResultDims, transposePerm);
+    }
+
+    // Step 3: Calculate the final packed result dimensions through the inverse
+    // result dimensions permutation map. This effectively linearizes the packed
+    // result dimensions with respect to the output dimensions. For example, if
+    // the permuted output dimensions are [D0, D2, D1], this will transform all
+    // packed operand result dimensions with the permutation map that would make
+    // the output dimensions the identity map [D0, D1, D2], i.e. {D0 -> D0, D1
+    // -> D2, D2 -> D1}. Suppose that the operand dimensions are [D0, D2], this
+    // operation would transform it into [D0, D1] to align with the output
+    // identity map.
+    //
+    // Running example:
+    //
+    // The packed and swizzled result dimensions for the input operand:
+    //
+    // packedResultDims:      [0, 2, 7, 6]
+    //
+    // Now we need to account for swizzled output result dimensions being
+    // linearized to the identity map. This can be achieved by applying
+    // `invOutSwizzlePerm` ([0, 1, 2, 5, 3, 6, 7, 4]):
+    //
+    // finalPackedResultDims: [0, 2, 4, 7]
+    SmallVector<int64_t> finalPackedResultDims = llvm::map_to_vector(
+        packedResultDims, [&](int64_t r) { return invOutSwizzlePerm[r]; });
+
+    // Create the packed indexing map.
+    SmallVector<AffineExpr> packedResultExprs =
+        llvm::map_to_vector(finalPackedResultDims, [&](int64_t dim) {
+          return rewriter.getAffineDimExpr(dim);
+        });
+    auto packedInputMap = AffineMap::get(
+        /*dimCount=*/iteratorTypes.size(), /*symbolCount=*/0, packedResultExprs,
+        rewriter.getContext());
+    packedIndexingMaps.push_back(packedInputMap);
+  }
+  // Create the new packed identity map for the output.
+  packedIndexingMaps.append(
+      genericOp.getNumDpsInits(),
+      rewriter.getMultiDimIdentityMap(convertedResultType.getRank()));
+  SmallVector<Type> convertedResultTypes =
+      llvm::map_to_vector(genericOp.getResultTypes(), [&](Type t) -> Type {
+        return RankedTensorType::get(
+            convertedResultType.getShape(),
+            cast<RankedTensorType>(t).getElementType());
+      });
+  auto materializedGenericOp = linalg::GenericOp::create(
+      rewriter, genericOp.getLoc(), convertedResultTypes,
+      convertedInputOperands, convertedOutputOperands, packedIndexingMaps,
+      iteratorTypes,
+      /*bodyBuild=*/nullptr, linalg::getPrunedAttributeList(genericOp));
+  rewriter.inlineRegionBefore(genericOp.getRegion(),
+                              materializedGenericOp.getRegion(),
+                              materializedGenericOp.getRegion().begin());
```

Comment:
We don't really need the `rewriter`. It can be something like:

```
genericOp.getRegion().cloneInto(&materializedGenericOp.getRegion(), mapper);
```

The change is mainly for showing the [diff](https://gist.github.com/hanhanW/e89aa9e5052f8db37b2543fa368ed605). I can remove the `rewriter` in a follow-up commit within this PR. Let me know what you think, thanks!